crates/biome_js_type_info/src/resolver.rs

```diff
@@ -1,54 +1,144 @@
+        // Clear the bits of the old module ID, while preserving the resolver
+        // level, and OR with the bits from the new module ID.
+        Self((self.0 & LEVEL_MASK) | module_id.0, self.1)
+    }
+}
+
+/// Identifier that indicates which module a type is defined in.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct ModuleId(u32);
+
+impl ModuleId {
+    pub const fn new(index: usize) -> Self {
+        // Top two bits are reserved to fit in resolver level.
+        debug_assert!(index < MODULE_ID_MASK as usize);
+
+        Self(index as u32)
     }
 
-    pub const fn level(&self) -> TypeResolverLevel {
-        self.0
+    pub const fn index(self) -> usize {
+        self.0 as usize
     }
 }
 
 /// Indicates the level within which a symbol has been resolved.
 ///
-/// The level is used by type resolvers to determine whether to look up a given
-/// [`TypeId`] within their own level, or to forward the resolution to another
-/// resolver that can handle the level.
+/// The level is used by type resolvers to determine _where_ to look up a given
+/// [`TypeId`]. They can look up types within their own registered types, within
+/// modules they may have access to and/or decide to forward resolution to
+/// another resolver that may be able to handle the level.
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
 pub enum TypeResolverLevel {
-    /// Used for ad-hoc inference that is not cached anywhere.
+    /// Used for ad-hoc inference that is not cached except in the ad-hoc
+    /// resolver itself (which is discarded after use).
     AdHoc,
 
     /// Used for resolving types that exist within the same module as from which
     /// the resolution took place.
+    ///
+    /// A [`ResolvedTypeId`] that uses this level may have a [`ModuleId`] stored
+    /// as well. However, we **don't** store such module IDs as part of a
+    /// module's type information, because a module is unaware of its own ID.
+    /// Instead, we rely on the resolver to attach the module ID at resolution
+    /// time.
     Module,
 
     /// Used for resolving types that exist across modules within the project.
+    ///
+    /// Currently, we don't store resolved IDs with this level in the module
+    /// info. Instead, we use it to during a module's type collection to flag
+    /// resolved types that require imports from other modules. Such resolved
+    /// IDs then get converted to [`TypeReference::Imported`] before storing
+    /// them in the module info.
     Project,
 
     /// Used for language- and environment-level globals.
     Global,
 }
 
+impl TypeResolverLevel {
+    /// Creates `TypeResolverLevel` from the two least significant bits of a
+    /// `u32`.
+    ///
+    /// Only the two least significant bits may be set in order to let the type
+    /// fit into `ResolvedTypeId`. If more bits become necessary, we may need to
+    /// rebalance the layout of `ResolvedTypeId`.
+    pub const fn from_u2(bits: u32) -> Self {
```

Comment:
`from_u32`

Reply:
I actually named it like that on purpose, to make people conscious of the size restraint, even if we don't have a real `u2` type :)